### PR TITLE
Enable ccache and Sphinx cache on Travis

### PR DESCRIPTION
**Pull request by jwiggins**
This provides a small speedup for subsequent builds of a pull request (after a successful first build).

### CODE REVIEW

File: skimage/_build.py [hunk 1]
```diff
@@ -24,6 +24,11 @@ def cython(pyx_files, working_path=''):
     if len(sys.argv) >= 2 and sys.argv[1] == 'clean':
         return
 
+    cython_options = {}
+    if os.env.get('CACHE_CYTHON', 'False').lower() == 'true':
```
Comment (blink1073) [inline skimage/_build.py]: should be `os.environ`

Comment (jwiggins) [inline skimage/_build.py]: Yeah... that's what I get for not smoke testing before pushing

File: tools/travis_script.sh [hunk 2]
```diff
@@ -9,7 +9,7 @@ section_end "Test.with.min.requirements"
 
 section "Build.docs"
 if [[ ($PY != 2.6) && ($PY != 3.2) ]]; then
-    make html
+    SPHINXBUILD="python $(which sphinx-build) -d $HOME/.cache/sphinx" make html
```
Comment (blink1073) [inline tools/travis_script.sh]: I think this needs to be: `export SPHINXBUILD="python $(which sphinx-build) -d $HOME/.cache/sphinx"; make html`

File: skimage/_build.py [hunk 3]
```diff
@@ -24,6 +24,11 @@ def cython(pyx_files, working_path=''):
     if len(sys.argv) >= 2 and sys.argv[1] == 'clean':
         return
 
+    cython_options = {}
+    if os.environ.get('CACHE_CYTHON', 'False').lower() == 'true':
```
Comment (blink1073) [inline skimage/_build.py]: I tried doing this on my local machine, and it seems like caching is does not work for older versions of Cython, and appears to be broken in Python 3, due to this line: https://github.com/cython/cython/blob/master/Cython/Build/Dependencies.py#L542.

Comment (blink1073) [inline skimage/_build.py]: I'd say just punt on this one.

Comment (jwiggins) [inline skimage/_build.py]: Hmm... OK. `ccache` doesn't seem to be having an affect either. I was hoping that the Cython caching might help.

**Comment by stefanv** [comment]
Nice work!

**Comment by jwiggins** [comment]
How irritating...

File: tools/travis_script.sh [hunk 4]
```diff
@@ -9,7 +9,10 @@ section_end "Test.with.min.requirements"
 
 section "Build.docs"
 if [[ ($PY != 2.6) && ($PY != 3.2) ]]; then
-    make html
+    SPHINXBUILD="python $(which sphinx-build)"
```
Comment (blink1073) [inline tools/travis_script.sh]: `$(which sphinx-build)` -> `` `which sphinx build` ``

Comment (jwiggins) [inline tools/travis_script.sh]: abd0f48 used backticks and got a similar result as dc394ab. Am I missing something obvious?

Comment (blink1073) [inline tools/travis_script.sh]: Ha, yep!  This line: `pip install -q sphinx`

Comment (blink1073) [inline tools/travis_script.sh]: It is part of the top level Makefile, we overlooked it.

Comment (jwiggins) [inline tools/travis_script.sh]: Do you want to remove it from that Makefile?

Comment (blink1073) [inline tools/travis_script.sh]: No, it is still useful for local builds.

Comment (blink1073) [inline tools/travis_script.sh]: You can add `sphinx` to optional_requirements.txt

Comment (blink1073) [inline tools/travis_script.sh]: Then you'd have to move the docs build to after the optional installs.

**Comment by blink1073** [comment]
Argh, how about we use `doc/Makefile`, but allow overrriding the cache dir?:

```diff
diff --git a/doc/Makefile b/doc/Makefile
index 82478f0..bebb8e0 100644
--- a/doc/Makefile
+++ b/doc/Makefile
@@ -6,11 +6,12 @@ PYTHON        ?= python
 SPHINXOPTS    ?=
 SPHINXBUILD   ?= python $(shell which sphinx-build)
 PAPER         ?=
+SPHINXCACHE   ?= build/doctrees

 # Internal variables.
 PAPEROPT_a4     = -D latex_paper_size=a4
 PAPEROPT_letter = -D latex_paper_size=letter
-ALLSPHINXOPTS   = -d build/doctrees $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) source
+ALLSPHINXOPTS   = -d $(SPHINXCACHE) $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) source
 DEST = build
 .PHONY: all help clean html dirhtml pickle json htmlhelp qthelp latex changes linkcheck doctest gitwash gh-pages release_notes random_gallery
```

**Comment by jwiggins** [comment]
Spinx caching keeps failing with the following error:
```
Running Sphinx v1.2.3
/home/travis/venv/lib/python3.4/site-packages/sphinx/util/pycompat.py:83: DeprecationWarning: 'U' mode is deprecated
  f = open(filepath, 'rbU')
/home/travis/build/scikit-image/scikit-image/doc/source/conf.py:78: ResourceWarning: unclosed file <_io.TextIOWrapper name='../../skimage/__init__.py' mode='r' encoding='UTF-8'>
  setup_lines = open('../../skimage/__init__.py').readlines()
loading pickled environment... failed: [Errno 2] No such file or directory: '/home/travis/.cache/sphinx/environment.pickle'
Exception occurred:
  File "/home/travis/venv/lib/python3.4/site-packages/sphinx/ext/intersphinx.py", line 188, in load_mappings
    app.info('loading intersphinx inventory from %s...' % inv)
TypeError: not all arguments converted during string formatting
```
It failed in that was on a previous run as well.

File: tools/travis_script.sh [hunk 5]
```diff
@@ -55,6 +49,14 @@ pip list
 section_end "Install.optional.dependencies"
 
 
+section "Build.docs"
+if [[ ($PY != 2.6) && ($PY != 3.2) ]]; then
+    export SPHINXCACHE=$HOME/.cache/sphinx
```
Comment (blink1073) [inline tools/travis_script.sh]: I tried putting these on the same line and it worked.  Isn't bash fun?

https://travis-ci.org/blink1073/scikit-image/jobs/79185050

Comment (jwiggins) [inline tools/travis_script.sh]: I'm not sure why that should work, but I'm giving it a try now.

**Comment by jwiggins** [comment]
Still seems to fail during the intersphinx initialization. Ahhh... The sphinx install isn't getting a new enough version, so the `intersphinx_mapping` isn't being parsed correctly.

**Comment by jwiggins** [comment]
Finally... Gah.

**Comment by blink1073** [comment]
:clap: thanks for hanging with it.

**Pull request merged**